libp2p/autotls/acme/api.nim

```diff
@@ -0,0 +1,387 @@
+    )
+
+  if challengeResponse.authorizations.len() == 0:
+    raise newException(ACMEError, "Authorizations field is empty")
+
+  # get challenges
+  let authorizationsResponse = handleError("requestChallenge (authorizations)"):
+    (
+      await (
+        await HttpClientRequestRef
+        .get(self.session, challengeResponse.authorizations[0])
+        .get()
+        .send()
+      ).getResponseBody()
+    ).to(ACMEAuthorizationsResponse)
+
+  let dns01 = authorizationsResponse.challenges.filterIt(it.`type` == "dns-01")[0]
+
+  return ACMEChallengeResponseWrapper(
+    finalizeURL: challengeResponse.finalize, orderURL: orderURL, dns01: dns01
+  )
+
+proc notifyChallengeCompleted*(
+    self: ACMEApi,
+    chalURL: string,
+    key: KeyPair,
+    kid: Kid,
+    retries: int = DefaultChalCompletedRetries,
+): Future[void] {.async: (raises: [ACMEError, CancelledError]).} =
+  let emptyPayload = newJObject()
+
+  let completedResponse = handleError("notifyChallengeCompleted (notify)"):
+    (await self.signedAcmeRequest(chalURL, emptyPayload, key, kid = Opt.some(kid))).body.to(
+      ACMECompletedResponse
+    )
+
+  # check until acme server is done (poll validation)
+  for i in 0 .. retries:
+    let (retryAfterHeader, checkResponse) = handleError(
+      "notifyChallengeCompleted (check " & $i & ")"
+    ):
+      let rawResponse = await HttpClientRequestRef
+      .get(self.session, completedResponse.checkURL)
+      .get()
+      .send()
+      (
+        rawResponse.headers.getString("Retry-After"),
+        (await rawResponse.getResponseBody()).to(ACMECheckResponse),
+      )
+
+    case checkResponse.status
+    of "pending":
+      var retryAfter: Duration
+      try:
+        retryAfter = parseInt(retryAfterHeader).seconds
+      except ValueError:
+        retryAfter = DefaultChalCompletedRetryTime
+      await sleepAsync(retryAfter) # try again after some delay
+    of "valid":
+      return
+    else:
+      raise newException(
+        ACMEError,
+        "Failed challenge completion: expected 'valid', got '" & checkResponse.status &
+          "'",
+      )
+
+proc finalizeCertificate*(
+    self: ACMEApi,
+    domain: string,
+    finalizeURL: string,
+    orderURL: string,
+    key: KeyPair,
+    kid: Kid,
+    retries: int = DefaultFinalizeRetries,
+): Future[bool] {.async: (raises: [ACMEError, CancelledError]).} =
+  var certKey: cert_key_t
+  var certCtx: cert_context_t
+  var derCSR: ptr cert_buffer = nil
+
+  let personalizationStr = "libp2p_autotls"
+  if cert_init_drbg(
+    personalizationStr.cstring, personalizationStr.len.csize_t, certCtx.addr
+  ) != CERT_SUCCESS:
+    raise newException(ACMEError, "Failed to initialize certCtx")
+  if cert_generate_key(certCtx, certKey.addr) != CERT_SUCCESS:
+    raise newException(ACMEError, "Failed to generate cert key")
+
+  if cert_signing_req(domain.cstring, certKey, derCSR.addr) != CERT_SUCCESS:
+    raise newException(ACMEError, "Failed to create CSR")
```

Comment:
could this code be extracted to separate function? 

- it would make `finalizeCertificate` shorter, thus easier to read
- that utility function could be tested individually
- ....